src/Infrastructure/Data/EfRepository.cs

```diff
@@ -63,17 +66,19 @@ public async Task DeleteAsync(T entity)
 
         public async Task<T> FirstAsync(ISpecification<T> spec)
         {
-            return await ApplySpecification(spec).FirstAsync();
+            var specificationResult = await ApplySpecification(spec);
+            return await specificationResult.FirstAsync();
         }
 
         public async Task<T> FirstOrDefaultAsync(ISpecification<T> spec)
         {
-            return await ApplySpecification(spec).FirstOrDefaultAsync();
+            var specificationResult = await ApplySpecification(spec);
+            return await specificationResult.FirstOrDefaultAsync();
         }
 
-        private IQueryable<T> ApplySpecification(ISpecification<T> spec)
+        private async Task<IQueryable<T>> ApplySpecification(ISpecification<T> spec)
         {
-            return SpecificationEvaluator<T>.GetQuery(_dbContext.Set<T>().AsQueryable(), spec);
+            return await EfSpecificationEvaluator<T>.GetQuery(_dbContext.Set<T>().AsQueryable(), spec);
```

Comment:
If `GetQuery` is async it should say so in its name for clarity.
Same for `ApplySpecification`.